src/ring_theory/derivations.lean

```diff
@@ -0,0 +1,200 @@
+def transitive_module (R : Type*) (A : Type*) (M : Type*)
+  [comm_semiring R] [semiring A] [algebra R A] [add_comm_monoid M] [semimodule A M] :
+  semimodule R M :=
+{ smul_add := λ r x y, smul_add _ _ _,
+  smul_zero := λ r, smul_zero _,
+  zero_smul := λ x, show algebra_map R A 0 • x = 0, by rw [map_zero, zero_smul],
+  one_smul := λ x, show algebra_map R A 1 • x = x, by rw [map_one, one_smul],
+  mul_smul := λ r s x, show algebra_map R A (r * s) • x =
+    algebra_map R A r • algebra_map R A s • x, by rw [map_mul, mul_smul],
+  add_smul := λ r s x, show algebra_map R A (r + s) • x =
+    algebra_map R A r • x + algebra_map R A s • x, by rw [map_add, add_smul],
+  .. transitive_scalar R A M }
+
+/-- This class register an instance that the scalar multiplication over `M` is indeed defined as
+above in `transitive_scalar`. -/
+@[protect_proj]
+class compatible_semimodule (R : Type*) (A : Type*) [comm_semiring R] [semiring A] [algebra R A]
+(M : Type*) [add_comm_monoid M] [semimodule A M] [semimodule R M] :=
+(compatible_smul (r : R) (m : M) : r • m = ((algebra_map R A) r) • m)
+
+instance algebra.to_compatible_semimodule  {R : Type*} {A : Type*} [comm_semiring R] [semiring A]
+  [algebra R A] : compatible_semimodule R A A :=
+⟨λ r a, by rw [smul_def]; refl⟩
+
+section compatible_semimodule
+
+variables {R : Type*} [comm_semiring R]
+(A : Type*) [comm_semiring A] [algebra R A]
+{M : Type*} [add_comm_monoid M] [semimodule A M] [semimodule R M] [compatible_semimodule R A M]
+{N : Type*} [add_comm_monoid N] [semimodule A N] [semimodule R N] [compatible_semimodule R A N]
+
+@[simp] lemma compatible_smul (r : R) (m : M) : r • m = ((algebra_map R A) r) • m :=
+compatible_semimodule.compatible_smul r m
+
+variable {A}
+
+@[simp] lemma compatible_smul_comm (r : R) (a : A) (m : M) : a • r • m = r • a • m :=
+by rw [compatible_smul A r m, compatible_smul A r (a • m), ←mul_smul, mul_comm, mul_smul]
+
+@[simp] lemma compatible_map_smul (f : M →ₗ[A] N) (r : R) (m : M) :
+  f (r • m) = r • f m :=
+by rw [compatible_smul A r m, linear_map.map_smul, ←compatible_smul A r (f m)]
+
+instance : has_coe (M →ₗ[A] N) (M →ₗ[R] N) :=
+⟨λ f, ⟨f.to_fun, λ x y, f.map_add' x y, λ r n, compatible_map_smul _ _ _⟩⟩
+
+end compatible_semimodule
+
+/-- Derivations in commutative algebra. To be changed.-/
+@[protect_proj]
+structure derivation (R : Type*) (A : Type*) [comm_semiring R] [comm_semiring A]
+  [algebra R A] (M : Type*) [add_cancel_comm_monoid M] [semimodule A M] [semimodule R M]
+  [compatible_semimodule R A M]
+  extends A →ₗ[R] M :=
+(leibniz' (a b : A) : to_fun (a * b) = a • to_fun b + b • to_fun a)
+
+namespace derivation
+
+section
+
+variables {R : Type*} [comm_semiring R]
+{A : Type*} [comm_semiring A] [algebra R A]
+{M : Type*} [add_cancel_comm_monoid M] [semimodule A M] [semimodule R M]
+[compatible_semimodule R A M]
+(D : derivation R A M) {D1 D2 : derivation R A M} (r : R) (a b : A)
```

Comment:
I prefer to have multiple  `variables` declarations if they don't fit on one line.
```suggestion
variables {R : Type*} [comm_semiring R]
variables {A : Type*} [comm_semiring A] [algebra R A]
variables {M : Type*} [add_cancel_comm_monoid M] [semimodule A M] [semimodule R M]
variables [compatible_semimodule R A M]
variables (D : derivation R A M) {D1 D2 : derivation R A M} (r : R) (a b : A)
```